### PR TITLE
Add 8-byte Transaction ID

### PR DESCRIPTION
**Pull request by morgando**
**Context**

This PR is part of a larger project to reimplement snapshot isolation without using logical logs. This implementation will use a structure that maps a transaction to its commit LSN. Using the existing 4-byte transaction IDs as keys in this structure would be problematic because these keys are not guaranteed to be unique. This motivated the changes in this PR, which add an 8-byte transaction ID that is guaranteed to be unique. This will not only contribute to the ongoing snapshot rewrite but will also support building other features that rely on uniquely identifying transactions. In the future, 8-byte transaction IDs can replace 4-byte IDs entirely.

In the rest of this PR, an 8-byte transaction ID is referred to as a utxnid (for "unique transaction id").

**Enabling this feature**

[`gbl_utxnid_log`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-91454b1e6e0271e4ab1900d4829d866b37f219f943ed9864fe0f2f23e4cc91e2R449) is the switch that toggles this feature. It can be explicitly modified with the tunable [`utxnid_log`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-c6ffbd9fb3d6b94c5143be665b399991db0791d8b00bf2c3209a3d210daf2c7aR2065). It is OFF by default in the Bloomberg environment (see [`legacy_options`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-7a4eec22b2e10f64fb8c50ecae0d7a0227ddbd692a6ff79679893f75dc381b7bR430)) but ON by default elsewhere.

**Implementation**

*Assigning utxnids*
When a transaction begins, a utxnid is assigned to it from [`next_utxnid`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-fb9f5c855be885e551d31c68861135fc7e41e6909a8b41b3e7ce6e5d7d870b5cR2736) (see [`__txn_begin_int_int`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-bf04707b376550c402355a1dc5ba543c3d459d99630b7c364cb66862874088a7R546)). During recovery, `next_utxnid` is updated so that it is greater than any utxnid already in the log (see [`__db_dispatch`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-3639dbf86938846cce9717b28a047ef478d5c8f35d9df519779ce3ae78bb2b1fR462)). 

*Versioning log records*
If this feature is enabled, then any logged record will include the utxnid of the associated transaction *and* will have 2000 added to its base rectype (see [`gen_rec.awk`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-da92cd605c6d4ff9983015e274574471ab873f18f61767145a2d2dfe3c78d91eR463)). Adding 2000 to the rectype serves to distinguish the version of the record. This versioning technique is the same technique that is used by ufid-logging, which adds 1000 to the base rectype. After these changes, a record that does not include a ufid or a utxnid will have a rectype equal to its base rectype, a record that includes a ufid but not a utxnid will have a rectype equal to its base rectype plus 1000, a record that includes a utxnid but not a ufid will have a rectype equal to its base rectype plus 2000, and a record that includes both a ufid and a utxnid will have a rectype equal to its base rectype plus 3000.

There are points at which a rectype is extracted from a raw log record and is then evaluated. At each of these points, the extracted rectype is now normalized before it is evaluated using the function [`normalize_rectype`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-ec57e659bd52da1fac41087fbe2907e44ad664e1c593477769caa537a482a07fR392). If 2000 has been added to a rectype, then this function subtracts 2000 from a rectype and returns 1; otherwise, it does nothing and returns 0. It does not subtract 1000 from a rectype if 1000 has been added because this is already done by existing code where it is appropriate. If log records are versioned further in the future, then `normalize_rectype` may be extended to handle these versions as well.

**Testing**

This PR adds `utxnid.test` (see [`tests/utxnid.test/runit`](https://github.com/bloomberg/comdb2/pull/3712/files#diff-3d2e62eca8483e057100814e799a38c498d537e14d5616678362e9a62c1967c5R69)). This test suite verifies that every log record with a nonzero 4-byte transaction ID also has a nonzero utxnid that is unique to it. It also verifies that any utxnid assigned after a bounce or a master swing is greater than the largest utxnid in the log before this action occurs.

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ berkdb/txn/txn_auto.c @@
+ Please remove debugging statements, and replace fprintf/printf with logmsg.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>4/468 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_transactional_rowlocks_generated<br>sc_addfield<br>pmux_sqlite_file_generated<br>tunables</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>9/472 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>truncatesc<br>truncatesc_offline_generated<br>truncatesc_ufid_on_generated<br>physrep_ignore_table<br>ddl_analyze<br>pmux_sqlite_file_generated<br>tunables<br>phys_rep<br>nophyswrite</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>2/472 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>ddl_analyze<br>tunables</p>

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>0/473 tests failed</b> &#x26A0;.</p>

File: berkdb/rep/rep_record.c [hunk 1]
```diff
@@ -388,6 +389,17 @@ lc_free(DB_ENV *dbenv, struct __recovery_processor *rp, LSN_COLLECTION * lc)
 	lc->nalloc = 0;
 }
 
+int normalize_rectype(u_int32_t *rectype) {
+	if (*rectype > 12000 || (*rectype > 2000 && *rectype < 10000)) {
+		assert (gbl_utxnid_log);
```
Comment (markhannum) [inline berkdb/rep/rep_record.c:394]: I think remove this assert- we may need to disable utxnid logs & restart a database

File: berkdb/txn/txn_auto.c [hunk 2]
```diff
@@ -78,6 +80,9 @@ __txn_regop_log_int(dbenv, txnid, ret_lsnp, ret_contextp, flags,
 #endif
 
 	rectype = DB___txn_regop;
```
Comment (markhannum) [inline berkdb/txn/txn_auto.c:82]: gbl_utxnid_log is toggle-able at runtime- copy the tunable at top & used the copied version

Comment (morgando) [reply berkdb/txn/txn_auto.c:83]: [Done here](https://github.com/bloomberg/comdb2/pull/3712/commits/2669ea1c2e1d2464f893eb2bfd8c4053b8538066).

**Review by markhannum — DISMISSED** [review]
Awesome

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Error &#x26A0;.<br>Cbuild submission: Error &#x26A0;.<br>Regression testing: <b>4/473 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>sc_truncate<br>phys_rep<br>sc_timepart<br>sc_timepart_logicalsc_generated</p>

**Review by markhannum — APPROVED** [review]

**Pull request merged**